core/shared/src/main/scala/zio/stm/TRandom.scala

```diff
@@ -0,0 +1,368 @@
+            }
+          }
+          .commit
+      }
+    }
+
+  private[zio] def nextDoubleBetweenWith(minInclusive: Double, maxExclusive: Double)(
+    nextDouble: STM[Nothing, Double]
+  ): STM[Nothing, Double] =
+    if (minInclusive >= maxExclusive)
+      STM.die(new IllegalArgumentException("invalid bounds"))
+    else
+      nextDouble.map { n =>
+        val result = n * (maxExclusive - minInclusive) + minInclusive
+        if (result < maxExclusive) result
+        else Math.nextAfter(maxExclusive, Float.NegativeInfinity)
+      }
+
+  private[zio] def nextFloatBetweenWith(minInclusive: Float, maxExclusive: Float)(
+    nextFloat: STM[Nothing, Float]
+  ): STM[Nothing, Float] =
+    if (minInclusive >= maxExclusive)
+      STM.die(new IllegalArgumentException("invalid bounds"))
+    else
+      nextFloat.map { n =>
+        val result = n * (maxExclusive - minInclusive) + minInclusive
+        if (result < maxExclusive) result
+        else Math.nextAfter(maxExclusive, Float.NegativeInfinity)
+      }
+
+  private[zio] def nextIntBetweenWith(
+    minInclusive: Int,
+    maxExclusive: Int
+  )(nextInt: STM[Nothing, Int], nextIntBounded: Int => STM[Nothing, Int]): STM[Nothing, Int] =
+    if (minInclusive >= maxExclusive) {
+      STM.die(new IllegalArgumentException("invalid bounds"))
+    } else {
+      val difference = maxExclusive - minInclusive
+      if (difference > 0) nextIntBounded(difference).map(_ + minInclusive)
+      else nextInt.repeatUntil(n => minInclusive <= n && n < maxExclusive)
+    }
+
+  private[zio] def nextLongBetweenWith(
+    minInclusive: Long,
+    maxExclusive: Long
+  )(nextLong: STM[Nothing, Long], nextLongBounded: Long => STM[Nothing, Long]): STM[Nothing, Long] =
+    if (minInclusive >= maxExclusive)
+      STM.die(new IllegalArgumentException("invalid bounds"))
+    else {
+      val difference = maxExclusive - minInclusive
+      if (difference > 0) nextLongBounded(difference).map(_ + minInclusive)
+      else nextLong.repeatUntil(n => minInclusive <= n && n < maxExclusive)
+    }
+
+  private[zio] def nextLongBoundedWith(n: Long)(nextLong: STM[Nothing, Long]): STM[Nothing, Long] =
+    if (n <= 0)
+      STM.die(new IllegalArgumentException("n must be positive"))
+    else {
+      nextLong.flatMap { r =>
+        val m = n - 1
+        if ((n & m) == 0L)
+          STM.succeedNow(r & m)
+        else {
+          def loop(u: Long): STM[Nothing, Long] =
+            if (u + m - u % m < 0L) nextLong.flatMap(r => loop(r >>> 1))
+            else STM.succeedNow(u % n)
+          loop(r >>> 1)
+        }
+      }
+    }
+
+  private[zio] def shuffleWith[A, Collection[+Element] <: Iterable[Element]](
+    nextIntBounded: Int => STM[Nothing, Int],
+    collection: Collection[A]
+  )(implicit bf: BuildFrom[Collection[A], A, Collection[A]]): STM[Nothing, Collection[A]] =
+    for {
+      bufferRef <- TRef.make(new scala.collection.mutable.ArrayBuffer[A])
```

Comment:
Refs shouldn't hold mutables.

Reply:
good catch! Blindly copied that from the random one

Reply:
Definitely a separate issue but we shouldn't be doing that in the original version either.